tests/Unit/Css/CssDocumentTest.php

```diff
@@ -437,15 +441,52 @@ public function provideCssWithoutNonConditionalAtRules(): array
      * @dataProvider provideValidAtCharsetRules
      * @dataProvider provideInvalidAtCharsetRules
      */
-    public function discardsValidOrInvalidAtCharsetRule(string $css): void
+    public function discardsValidOrInvalidAtCharsetRuleNotInDebugMode(string $css): void
     {
-        $subject = new CssDocument($css);
+        $subject = new CssDocument($css, false);
 
         $result = $subject->renderNonConditionalAtRules();
 
         self::assertSame('', $result);
     }
 
+    /**
+     * @test
+     *
+     * @param string $css
+     *
+     * @dataProvider provideValidAtCharsetRules
+     */
+    public function discardsValidAtCharsetRuleInDebugMode(string $css): void
+    {
+        $subject = self::createDebugSubject($css);
+
+        $result = $subject->renderNonConditionalAtRules();
+
+        self::assertSame('', $result);
+    }
+
+    /**
+     * @test
+     *
+     * @param string $css
+     *
+     * @dataProvider provideInvalidAtCharsetRules
+     */
+    public function throwsExceptionForOrDiscardsInvalidAtCharsetRuleInDebugMode(string $css): void
+    {
+        try {
+            $subject = self::createDebugSubject($css);
+
+            $result = $subject->renderNonConditionalAtRules();
+
+            self::assertSame('', $result);
+        } catch (UnexpectedEOFException $exception) {
```

Comment:
Please let's use `$this->expectException`, `expectExceptionCode` and `expectExceptionMessage` instead.

Ditto for other tests that check for exceptions.

Reply:
See above.

Reply:
I mean below.

Reply:
> Please let's use `$this->expectException`, `expectExceptionCode` and `expectExceptionMessage` instead.

Though if there is a `doNotExpectException` method the test could be written without the `try`/`catch` - i.e. we expect an exception but if we reach some later point we'll make an assertion instead, and cancel the expectation of the exception.  Is that possible?

Reply:
> Though if there is a `doNotExpectException` method the test could be written without the `try`/`catch` ... Is that possible?

`expectException(null)` would do the trick, except that the parameter is not documented as nullable and a Psalm error is generated.

Reply:
To test that an exception is _not_ thrown, we can do one of the following things:

a) state that not throwing the exception is the main statement of the test, and use a `@doesNotPerformAssertions` annotation
b) state that the return value (or something similar) is the main statement of the test, and assert for that

Reply:
But we can't, it seems, have a *'throws exception'* ***OR*** *'satisfies assertion'* test.  Or, at least, not easily.

Reply:
Then please let's test for one of the two. (We can also add the other case as a test and mark it as `self::markAsSkipped()` (if I remember the method name correctly).

Reply:
OK.  I've change this (and the other similar test) to expect an exception, but if one isn't thrown mark the test as skipped - since in strict (debug) mode the CSS parser only throws an exception with some of the datasets, and we are already testing that the invalid rules from these datasets are dropped in lenient (non-debug) mode.